core/src/main/java/org/apache/gravitino/storage/relational/JDBCBackend.java

```diff
@@ -569,6 +593,103 @@ public <E extends Entity & HasIdentifier> List<E> updateEntityRelations(
     }
   }
 
+  @Override
+  public int deleteEntitiesAndRelations(
+      Type relType, Relation.VertexType deleteVertexType, List<Relation> relations)
+      throws IOException {
+    switch (relType) {
+      case METADATA_OBJECT_STAT_REL:
+        Preconditions.checkArgument(
+            deleteVertexType == Relation.VertexType.DESTINATION,
+            "Only supports to delete relations of destination vertex");
+
+        AtomicInteger deleteCount = new AtomicInteger(0);
+        Map<Pair<NameIdentifier, Entity.EntityType>, List<Relation>> groupedRelations =
+            relations.stream().collect(Collectors.groupingBy(Relation::getSourceVertex));
+
+        if (groupedRelations.isEmpty()) {
+          return 0; // No relations to delete.
+        }
+
+        Preconditions.checkArgument(
+            groupedRelations.size() == 1,
+            "The relations must have the same source identifier and type, but got %s groups",
+            groupedRelations.size());
+
+        groupedRelations.forEach(
```

Comment:
If `groupedRelations` size is always 1, don't need to use loop.

Reply:
OK, I modified this place.